File: 3. Behavioral Design Patterns/Command/Java/CommandTemplate.java
// Command Pattern – Java Template

interface Command { void execute(); void undo(); }

class Receiver { void action(){ /* do something */ } void undoAction(){ /* undo something */ } }

class ConcreteCommand implements Command {
    private final Receiver r;
    ConcreteCommand(Receiver r){ this.r = r; }
    public void execute(){ r.action(); }
    public void undo(){ r.undoAction(); }
}

class Invoker { private Command cmd; void set(Command c){ cmd = c; } void run(){ if (cmd != null) cmd.execute(); } void undo(){ if (cmd != null) cmd.undo(); } }

public class CommandTemplate { public static void main(String[] args){ Invoker i = new Invoker(); i.set(new ConcreteCommand(new Receiver())); i.run(); i.undo(); } }

File: 3. Behavioral Design Patterns/Command/C#/CommandTemplate.cs
// Command Pattern – C# Template

namespace CommandTemplate
{
    interface ICommand { void Execute(); void Undo(); }

    class Receiver { public void Action(){ /* do something */ } public void UndoAction(){ /* undo something */ } }

    class ConcreteCommand : ICommand
    {
        private readonly Receiver _r;
        public ConcreteCommand(Receiver r){ _r = r; }
        public void Execute(){ _r.Action(); }
        public void Undo(){ _r.UndoAction(); }
    }

    class Invoker { private ICommand _cmd; public void Set(ICommand c){ _cmd = c; } public void Run(){ _cmd?.Execute(); } public void Undo(){ _cmd?.Undo(); } }

    class Program { static void Main(){ var r = new Receiver(); var c = new ConcreteCommand(r); var inv = new Invoker(); inv.Set(c); inv.Run(); inv.Undo(); } }
}
